### ISSUE TITLE
getting-started rethink

### ISSUE THREAD

**Issue by AdrianRossouw** [senecajs/getting-started#8]
I've been processing what about this current tutorial just doesn't sit right with me, and i've started to think that the underlying issue is really that the majority of what us developers do on the web is really pretty REST and crud-oriented.

Therefor, I think that the best way to explain seneca to people is to use something closer to what they know (ie: a CRUD site).Where we run into a roadblock here, is that seneca has a rather large subsystem called 'entities' that are supposed to automagically take care of all this stuff for you.  

But a getting-started tutorial really shouldn't just point you at this black box and assume you won't hit the pitfalls they represent.

The reaction to this, is of course to make the examples incredibly abstract. This is why this current tutorial has half a dozen math-*.js files instead of anything relatable.
### So what I think
1. We should write a new tutorial that is a simple CRUD app.  
   I like using a pet shop as an example, because lolcats. (also: Petting Started tutorial... haha)
2. It should be a simple HAPI (or express) site.  
   Use server side rendering, avoid the whole react vs. whatever fight.
3. It should **not** use anything other than base seneca.  
   This includes not using seneca entities. We can link entities at the end of the tutorial.
4. It should show off the pattern matching in a clean crud oriented way.  
5. later ... It can bring in multiple data sources hidden behind patterns  
   (ie: all dogs are loaded from the american dog adopter society's website instead of our own db)
6. at the end, It should probably bring in some patterns via transports, to show how running systems can get altered.

any thoughts ? 
@mcdonnelldean @codewriterwriter @sjmcdowall @bastianwegge

**Comment by mcdonnelldean** [comment]
I'm happy enough for this to change. The current version is not working for the community. Our litmus test needs to be how it is perceived by those it targets.

I'm also happy with this being a small little app with the notes above. I think it will be easier to understand if it is a cohesive piece rather than a set of random examples.

A couple of things,
1. There needs to be tests to ensure regression, this is killing us right now.
2. Entities need to be included (mem entities) unless we use no storage.

On entities, basically, if you can build it without them fine but if you need to persist we need to use entities and not third party storage. My feeling here is that it shouldn't use entities but rather include entities as a `next steps` portion of the guide to allow folks to extend the getting started sample.

**Comment by bastianwegge** [comment]
Yes of course! I like to relate to the ActiveRecord in Ruby or EntityFramework in .NET as references on how to start this tutorial.

My ideal tutorial would look something like this:
1. Basic logic like sum/product for the start and to explain how to use the `add` and `act` methods.
2. Separate into files, where just pointing out that `seneca.use('math')` ist the best option to go. Maybe some more explanation about what's going on there exactly.
3. CRUD-Actions: When it comes to developers, I think that most people finding seneca are searching for "nodejs microservices". If you think about it, then the front-end part almost immediately turns out to be trivial. Sure let's use `mongodb` or `mongoose` since that's what the majority is using.
4. Introduction into seneca entities: Since everybody wants to be database-independent but most of us are using mongoose since we have no alternative, this could really be a game-changer. You're already profiting out of the system and continue to use it, so why not publish some parts of it? We could just explain what to do in terms of migrating from mongoose to entities and maybe give some examples about hooks that can be done using the `seneca.wrap` function.
5. splitting web & api: Split in terms of location. Explain the `listen` function with its arguments and how to best connect them based on some field-experience.

We need to think about what's really bothering "right now" and then form a tutorial out of that. Things like logging, monitoring, service-discovery, Mesh / SWIM should be part of an advanced tutorial or at least the "real" documentation. I'm really interested in hearing about your logging experience with seneca, but right now I'd love to just understand what I have to do to my systems in order to get up and running fast.

**Comment by AdrianRossouw** [comment]
the issue with diving into seneca entities straight away is that they try to abstract away all of the method passing and actions away from your code. so you'd end up with a tutorial that never uses anything other than the .save / .load / .make, which doesn't equip you for using seneca itself.

I was going to suggest just using in-memory data as a way to store data.

```

var pets = {
    john: { type: 'fish', color: 'gold', age: 12 },
    susan: { type: 'cat', color: 'white', age: 2}
}

module.exports = function(config) {
   var seneca = this;

   seneca.add('role:pets,cmd:create', function(data, done) {
        if(pets[data.name]) { return done('record already exists'); }
        pets[data.name] = data;
        done(null, pets[data.name]));
   });

   seneca.add('role:pets,cmd:remove', function(data, done) {
       if (!pets[data.name]) { return done('no such pet'); }
       delete pets[data.name];
       done();
   });

   seneca.add('role:pets,cmd:update', function(data, done) {
       if (!pets[data.name]) { return done('no such pet'); }
       pets[data.name] = data;
       done(null. pets[data.name]);
  });

  return {name: 'pet-store'};
})
```

and then have another layer added in for a specific type of pet (ie: type:dog) that will speak to an external api or something instead.

this would show how to register actions, how more specific actions win, and in the HAPI code how to execute actions.

At the end of this all I would have a block that you can also use seneca entities to do a lot of this stuff for you, with a link to a separate document explaining how to get up and running with entities. (which explains swappable stores, etc)

**Comment by mcdonnelldean** [comment]
Yeah that sort of thing 😉

**Comment by mcdonnelldean** [comment]
The most important point is that it is easy to wrok through and gives folk enough info that they could reasonably kick around something like nodezoo after a week of messing with it. Solid foundations I think is best

**Comment by bastianwegge** [comment]
@AdrianRossouw I cannot really follow your argumentation. If you say entities is one of the important parts (and who doesn't want this to be abstracted) then why not tackle it right away? We're driven by patterns that are explained to us that make sense. If we suggest that people to use entities and solve common problems like hooks, aggregation or data-optimization we might have a good chance for better acceptance.

Also what would be your ideal equip for using seneca?

The last thing to add: Would you mind a test-driven tutorial? Since this is really my butter and bread I'd like to help as much as I can to create cases or tests for this. Writing tests when doing server-side code is the best way to go. Almost all tutorials use "then curl this" or "open your browser and go to localhost:8000" but is that what we really to after all?

I'd love to take part in this, let's just start working and come up with some results!

**Comment by kaktus42** [comment]
my 2 cents as someone just stumbeling upon seneca when searching for a microservice framework, being annoyed by pomelo because of the non-english community.

I liked the tutorial very much - thanks for that!
I found in this tutorial a good introduction on how to implement my business logic as a micro service with seneca.
I don't think it should be substituted by a tutorial focusing on the entity/databse part. It is just one of many applications for a microservice.
However, an additional tutorial going into depth with DB abstraction can be nice, of course.
